doc/decisions/key_name.md

```diff
@@ -0,0 +1,39 @@
+# Key Name
+
+## Problem
+
+Often a Key _ argument is used when you just need a key name.
+This is because with a Key _ we know the name is valid and we get an unescaped name.
+Using a Key \* here makes the API a bit confusing.
+
+There could be a richer API for manipulating key names without relying on the escaped name (e.g. concatenating two full key names).
+With the current situation, all these functions would need to be part of the API for a key. Adding such functions to the key API is certainly not minimal.
+
+## Constraints
+
+## Assumptions
+
+- the key name is actually, in every implementation, a plain string
```

Comment:
Good question, you somewhere assumed there could be a structure needed.

Reply:
I don't think a structure is _needed_, but it is a trade-off. The plain string we have right now allows for very simple and quick sorting of keys. It is also fairly memory efficient. I don't think you could use any less memory without sharing memory between keys.

A more structured key name, or additional stored information (e.g. pointers to each of the parts) could (in theory) speed up certain operations. But like I said, this would mean increased memory cost.

I just think, that this assumption is not actually required for anything that is decided here. Therefore, we shouldn't make the assumption and leave the decision up to the implementation (memory layout problems notwithstanding)

Reply:
> The plain string we have right now allows for very simple and quick sorting of keys

Thank you for reminding a reason that will probably convince you: The memcmp we use for looking up keys is order of magnitudes faster then anything what we could have with a more structured key name.

I think we should forget about this idea of structuring the name and use unescaped names as much as possible for fast creation of keys.

Reply:
Yes, in the current situation (only C and Rust implementations) the flat byte buffer is definitely the best solution for the key name.